UXD-308 Refactor stories and added mdx for FormElement (#741)
acl-services/paprika: packages/FormElement/stories/FormElement.stories.mdx

import RadioExample from "./examples/RadioExample";
import DatePickerExample from "./examples/DatePickerExample";
import AccessibilityExample from "./examples/AccessibilityExample";
import ListBoxExample from "./examples/ListBoxExample";
import ButtonGroupExample from "./examples/ButtonGroupExample";
import NativeInputExample from "./examples/NativeInputExample";
import NestedExample from "./examples/NestedExample";
import HTMLExample from "./examples/HTMLExample";
import ComponentLabelExample from "./examples/ComponentLabelExample";
import FormElement from "../src/FormElement";

<Meta
  title={`${getStoryName("FormElement")}/Docs`}
  component={FormElement}
  parameters={{ options: { showPanel: false } }}
/>

<Description markdown={Readme} />

## More Examples

### Index

- <a href="#accessibility">Accessibility</a>
- <a href="#basic-input-components">Basic Input Components</a>

  - <a href="#textarea">Textarea</a>

  - <a href="#checkbox">Checkbox</a>

- <a href="#complex-input-components">Complex Input Components</a>

  - <a href="#component-label">Component Label</a>

  - <a href="#datepicker">DatePicker</a>

### Accessibility

> The `<FormElement />` has been designed with accessibility in mind. Users can make use of `<FormElement.Instructions />`, `<FormElement.Description />`, `<FormElement.Error />`, and `<FormElement.Help` to add instructions, guidance and feedback to users. Error messages, instructions, and descriptions can be associated to the input field by applying the supplied `ariaDescribedBy` parameter in the render function supplied to the `children` of the `<FormElement.Content>`.

Reviewer: ```suggestion
The `<FormElement />` has been designed with accessibility in mind. Users can make use of `<FormElement.Instructions />`, `<FormElement.Description />`, `<FormElement.Error />`, and `<FormElement.Help />` to add instructions, guidance and feedback to users. Error messages, instructions, and descriptions can be associated to the input field by applying the supplied `ariaDescribedBy` parameter in the render function supplied to the `children` of the `<FormElement.Content>`.
```

> The `<FormElement.Help />` will add an icon with a tooltip next to the label.

Reviewer: Let's move this to last, after the paragraph about the `label`.

```jsx
<FormElement label="Form Label">
  <FormElement.Help>
    Give me some help. <a href="wegalvanize.com">Learn more</a>.
  </FormElement.Help>
  <FormElement.Instructions>
    <span>
      Example text for extra panel for questionnaires. Example text for extra panel for questionnaires Example text for
      extra panel for questionnaires. Example text for extra panel for questionnaires
    </span>
  </FormElement.Instructions>
  <FormElement.Content>
    {({ idForLabel, ariaDescribedBy }) => (
      <Input id={idForLabel} placeholder="Form placeholder" aria-describedby={ariaDescribedBy} />
    )}
  </FormElement.Content>
  <FormElement.Description>
    <span>This is description text</span>
  </FormElement.Description>
  <FormElement.Error>There was an error</FormElement.Error>
</FormElement>
```

##### Result

<FormElementStory>
  <AccessibilityExample />
</FormElementStory>

### Basic Input Components

The `<FormElement>` can be used with different components to fulfill a variety of input needs. This can be achieved by including basic input components such as the `<Select>`, `<Textarea>`, `<Checkbox>`, etc. as `children` for the `<FormElement.Content>`

#### Textarea

```jsx

> <FormElement hasFieldSet label="Form Label">

Reviewer: ```suggestion
<FormElement label="Form Label">
```

  <FormElement.Content>
    <Textarea defaultValue="Hello world">
  </FormElement.Content>
</FormElement>
```

##### Result

<FormElementStory>

>   <FormElement hasFieldSet label="Form Label">

Reviewer: ```suggestion
  <FormElement label="Form Label">
```

    <FormElement.Help>
      Give me some help. <a href="wegalvanize.com">Learn more</a>.

>     </FormElement.Help>

Reviewer: ✂️

    <FormElement.Content>
      <Textarea defaultValue="Hello world" />
    </FormElement.Content>
  </FormElement>
</FormElementStory>

#### Checkbox

```jsx
const optionsArray = ["Black Panther", "Wonder Woman", "Spiderman", "The Incredibles", "Thor", <span>test</span>];

const getCheckboxOptions = ariaDescribedBy =>
  optionsArray.map(hero => (
    <Checkbox
      ariaDescribedBy={ariaDescribedBy}
      key={hero}
      onChange={() => {}}
      isDisabled={false}
      size={Checkbox.types.size.MEDIUM}
    >
      {hero}
    </Checkbox>
  ));

<FormElement hasFieldSet label="Form Label">
  <FormElement.Content>

>     {({ ariaDescribedBy: ariaDescribedByOuter }) => getCheckboxOptions(ariaDescribedByOuter)}

Reviewer: 🤔  While using a `map()` is probably the way consumers will really use this, I wonder if it's simpler to just dump in the actual children.  Like:

```jsx
<Checkbox ariaDescribedBy={ariaDescribedBy} onChange={() => {}}>Black Panther</Checkbox>
<Checkbox ariaDescribedBy={ariaDescribedBy} onChange={() => {}}>Spiderman</Checkbox>
<Checkbox ariaDescribedBy={ariaDescribedBy} onChange={() => {}}>Wonder Woman</Checkbox>
```

Then we can remove the array and function above and focus on the API.  
What do you think?

  </FormElement.Content>
</FormElement>;
```

##### Result

<FormElementStory>
  <CheckboxExample />
</FormElementStory>

### Complex Input Components

The `<FormElement />` can handle a variety of components to further increase the users indivudal needs. This can range from a `<DatePicker />` component to a `<ListBox />` component. The following code snippet is an example of the `ComponentLabel` this example allows users to use the `label` property of the `<FormElement/ />`. Users can set the label property to equal an entire component, in this particular example making use of the `<Heading />` component.

> #### Component Label

Reviewer: Actually, now that you've got these examples split out, I think it makes more sense to do the Component Label example as its own `###` level heading.  It's not really a "complex component" example, just another more advanced option.

```jsx
<FormElement
  hasRequiredLabel={false}
  label={
    <Heading level={5}>
      <strong>
        <i>Form Label</i>
      </strong>
    </Heading>
  }
>
  <FormElement.Content>
    {({ idForLabel, ariaDescribedBy }) => (
      <Input
        id={idForLabel}
        onChange={() => {}}
        placeholder="Form placeholder"
        aria-describedby={ariaDescribedBy}
        aria-required={false}
        hasError={Boolean(errorText.length)}

>         size={FormElement.types.size.MEDIUM}

Reviewer: Let's drop `placeholder`, `aria-required`, `hasError`, `size`, just to keep this to its essentials so we don't distract from what we're really demonstrating.

      />
    )}
  </FormElement.Content>
</FormElement>
```

##### Result

<FormElementStory>
  <ComponentLabelExample />
</FormElementStory>

#### DatePicker

```jsx

> const errorText = "";

Reviewer: ✂️

<FormElement label="Form Label">
  <FormElement.Content>
    {({ idForLabel, ariaDescribedBy }) => (

>       <DatePicker onError={() => {}} hasError={Boolean(errorText.length)} id={idForLabel} onChange={() => {}}>

Reviewer: ```suggestion
      <DatePicker id={idForLabel} onChange={() => {}}>
```

import RadioExample from "./examples/RadioExample";
import DatePickerExample from "./examples/DatePickerExample";
import AccessibilityExample from "./examples/AccessibilityExample";
import ListBoxExample from "./examples/ListBoxExample";
import ButtonGroupExample from "./examples/ButtonGroupExample";
import NativeInputExample from "./examples/NativeInputExample";
import NestedExample from "./examples/NestedExample";
import HTMLExample from "./examples/HTMLExample";
import ComponentLabelExample from "./examples/ComponentLabelExample";
import FormElement from "../src/FormElement";

<Meta
  title={`${getStoryName("FormElement")}/Docs`}
  component={FormElement}
  parameters={{ options: { showPanel: false } }}
/>

<Description markdown={Readme} />

## More Examples

### Index

- <a href="#accessibility">Accessibility</a>
- <a href="#alternate-layouts">Alternate Layout</a>

  - <a href="#checkboxes">Checkboxes</a>
  - <a href="#radios">Radios</a>
  - <a href="#date-picker">Date Picker</a>
  - <a href="#listbox">ListBox</a>
  - <a href="#buttongroup">ButtonGroup</a>
  - <a href="#nested-formelement">Nested FormElements</a>

### Accessibility

> The Formelement has been designed with accessibiklity in mind. By using `<FormElement.Instructions>` users can add instructions for the end user. Users can also make use of the `<FormElement.Description>` to add additional direction for consumers.

Reviewer: ```suggestion
The `<FormElement>` has been designed with accessibility in mind. By using `<FormElement.Instructions>` users can add instructions for the end user. Users can also make use of the `<FormElement.Description>` to add additional directions for consumers. Both of these, along with any error messages can be associated to the input field by applying the supplied `ariaDescribedBy` parameter in the render function supplied to the `children` of `<FormElement.Content>`. 

In addition, a `<label>` is rendered by the `<FormElement>` and is associated to the input field by the `idForLabel` parameter of the same `children` function. 
```

Reviewer: Oh, I just noticed there's also a `FormElement.Help` worth mentioning above.  And if you can make it easier to understand that I did, please do!

```jsx
<FormElement label="Form Label">
  <FormElement.Instructions>
    <span>
      Example text for extra panel for questionnaires. Example text for extra panel for questionnaires Example text for
      extra panel for questionnaires. Example text for extra panel for questionnaires
    </span>
  </FormElement.Instructions>
  <FormElement.Content>
    {({ idForLabel, ariaDescribedBy }) => (
      <Input
        id={idForLabel}
        onChange={handleChange}
        value={value}
        placeholder="Form placeholder"
        aria-describedby={ariaDescribedBy}
        hasError={Boolean(errorText.length)}
      />
    )}
  </FormElement.Content>
  <FormElement.Description>
    <span>This is description text</span>
  </FormElement.Description>
  <FormElement.Error>{errorText}</FormElement.Error>
  <FormElement.Help>
    Give me some help. <a href="wegalvanize.com">Learn more</a>.
  </FormElement.Help>
</FormElement>
```

##### Result

<FormElementStory>
  <AccessibilityExample />
</FormElementStory>

> ### Alternate Layouts

Reviewer: I don't think this is the right heading... actually, I'm not sure we need to nest the following use cases under a heading.  I know they were all embedded on an Alternate Layouts story before, but I much prefer the way you've broken them out individually.  Let's drop this heading and elevate Checkboxes, Radios, etc to `###` level.

The FormElement can be used in a variety of formats. These can include FormElements with checkboxes, radios, listbox's etc.

#### Checkboxes

```jsx
<FormElement hasFieldSet label="Form Label">
  <FormElement.Content>
    {({ ariaDescribedBy: ariaDescribedByOuter }) => getCheckboxOptions(ariaDescribedByOuter)}
  </FormElement.Content>
  <FormElement.Description>
    <span>Description of this field.</span>
  </FormElement.Description>
  <FormElement.Error>{errorText}</FormElement.Error>
  <FormElement.Help>
    Give me some help. <a href="wegalvanize.com">Learn more</a>.
  </FormElement.Help>
</FormElement>
```

##### Result

<CheckboxExample />

#### Radios

```jsx
<FormElement hasFieldSet label="Form Label">
  <FormElement.Content>
    {({ ariaDescribedBy: ariaDescribedByOuter }) => getRadioOptions(ariaDescribedByOuter)}
  </FormElement.Content>
  <FormElement.Description>
    <span>Description of this field.</span>
  </FormElement.Description>
  <FormElement.Error>{errorText}</FormElement.Error>
  <FormElement.Help>
    Give me some help. <a href="wegalvanize.com">Learn more</a>.
  </FormElement.Help>
</FormElement>

> ```

Reviewer: This is basically the same pattern as the checkboxes.  Maybe we don't need the code snippet, we can just _explain_ that they're basically the same and show both results.  Let's keep them as separate Examples stories though, I like that... but for the DOCS page, maybe less is more (and less code to maintain if things change).

##### Result

<RadioExample />

> #### Date Picker

Reviewer: Similarly, I think we can have an example for more complex input components (DatePicker, ListBox, ButtonGroup, etc).  Let's explain that, show one code snippet (you can pick which one you think is the most helpful) and then maybe show all the results... or have some links to the other example stories (or maybe just the example story source code on GitHub).

  </FormElement.Description>
  <FormElement.Error>{errorText}</FormElement.Error>
  <FormElement.Help>
    Give me some help. <a href="wegalvanize.com">Learn more</a>.
  </FormElement.Help>
</FormElement>
```

##### Result

<DatePickerExample />

#### ListBox

```jsx
<FormElement label="Form Label">
  <FormElement.Content>{({ refLabel }) => <ListBox refLabel={refLabel}>{listboxOptions}</ListBox>}</FormElement.Content>
  <FormElement.Description>
    <span>Description of this field.</span>
  </FormElement.Description>
  <FormElement.Help>
    Give me some help. <a href="wegalvanize.com">Learn more</a>.
  </FormElement.Help>
</FormElement>
```

##### Result

<ListBoxExample />

#### ButtonGroup

```jsx
<FormElement label="Form Label" onClickLabel={() => refButtonGroup.current.focus()}>
  <FormElement.Content>
    {() => <ButtonGroup ref={refButtonGroup}>{buttonGroupOptions}</ButtonGroup>}
  </FormElement.Content>
  <FormElement.Description>
    <span>Description of this field.</span>
  </FormElement.Description>
  <FormElement.Help>
    Give me some help. <a href="wegalvanize.com">Learn more</a>.
  </FormElement.Help>
</FormElement>
```

##### Result

<ButtonGroupExample />

> #### Nested FormElement

Reviewer: This example has a lot of code, let's just explain that it's possible to nest the FormElement in another FormElement, why you might want to do that, and then link to the NestedExample story or github source.

Reviewer: While I don't think this Nested example is a common enough use case to include in the Docs, I do think that the `ComponentLabel` _is_ a useful example to include here.  Can you include it (with a brief 1-2 sentence explanation before the code block example?)

##### Result

<FormElementStory>
  <Tagline>Form Element with instructions component.</Tagline>
  <br />
  <AccessibilityExample />
</FormElementStory>

> ### Components

Reviewer: ```suggestion
### Basic Input Components
```

> The `<FormElement />` can be used with differant components to create specific input needs. Some of these can include components such as the checkbox or radio.

Reviewer: ```suggestion
The `<FormElement>` can be used with different components to fulfill a variety of input needs. This can include using basic input components such as the `<Select>`, `<Textarea>`, `<Checkbox>`,  etc.
```

Reviewer 2: I've preferred `<FormElement />` because `<FormElement>` isn't valid JSX but not a hill to die on 😄

##### Result

<FormElementStory>
  <Tagline>Form Element with checkboxes.</Tagline>
  <br />
  <CheckboxExample />
</FormElementStory>
<Rule />
<FormElementStory>
  <Tagline>Form Element with radios.</Tagline>
  <br />
  <RadioExample />

> </FormElementStory>

Reviewer: I have a few more ideas about this, but I think it's better to discuss together.

> ### Complex Components

Reviewer: ```suggestion
### Complex Input Components
```
